File: app/lib/shared/validation/PlanValidator.ts
import { Plan } from "../../domain/Plan";
import { profileValidator } from "./ProfileValidator";

export class PlanValidator {
  public validatePlan(plan: Omit<Plan, "id">): void {
    this.validatePlanTitle(plan);
    this.validatePlanProfile(plan);
  }

  public validatePlanProfile(plan: Omit<Plan, "id">): void {
    profileValidator.validateProfile(plan.profile);
  }

  public validatePlanTitle(plan: Omit<Plan, "id">): void {
    if (typeof plan.title !== "string") {
      throw new Error("Title must be a string");
    }
    if (!/^(?! )[A-Za-z0-9]+(?: [A-Za-z0-9]+)*$/.test(plan.title)) {
      throw new Error("Invalid characters in the title");
    }
    if (plan.title.length > 100) {
      throw new Error("Plan name must be ≤ 100 characters");
    }
  }
}
export let planValidator = new PlanValidator();

File: backend/app/core/validation/PlanValidator.py
import re

from app.core.domain import Plan
from app.core.validation.ProfileValidator import profile_validator


class PlanValidator:
    def validate_plan(self, plan: Plan) -> None:
        self.validate_plan_title(plan)
        self.validate_plan_profile(plan)

    def validate_plan_profile(self, plan: Plan) -> None:
        profile_validator.validate_profile(plan.profile)

    def validate_plan_title(self, plan: Plan) -> None:
        if not isinstance(plan.title, str):
            raise ValueError("Title must be a string")

        if not re.match(r"^(?! )[A-Za-z0-9]+(?: [A-Za-z0-9]+)*$", plan.title):
            raise ValueError("Title contains invalid characters")

        if len(plan.title) > 100:
            raise ValueError("Plan title must be <= 100 characters")


plan_validator = PlanValidator()
